bookstore/archive.py

```diff
@@ -29,6 +45,46 @@ def __init__(self, *args, **kwargs):
             s3_additional_kwargs={},
         )
 
+        self._thread_pool = ThreadPoolExecutor(max_workers=30)
+
+        self.locks: Dict[str, Lock] = {}
+
+    async def archive(self, record: ArchiveRecord):
+        """Process a storage write, only allowing one write at a time to each path
+        """
+        lock = self.locks.get(record.filepath)
+
+        try:
+            if lock is None:
+                lock = Lock()
+                self.locks[record.filepath] = lock
+
+            # Try to acquire the lock, and if it fails don't bother archiving
+            # as we should finish a previous write
+            await lock.acquire(1)
```

Comment:
I would recommend putting this whole lock in a with block.  If anything would be written before the try block below, an exception would cause the lock to never be released.

Reply:
Yeah, I could refactor to do that. I was trying to separate `tornado`'s `TimeoutError` (desired behavior when there's an in-progress write) out here with the steps that shouldn't fail.